### DISCUSSION TITLE
RFT mode circular import issue

### DISCUSSION THREAD

**Discussion by chendu1993** [dashingsoft/pyarmor#1842] [Q&A] [answered]
When running pyarmor with rft mode enabled the obfuscation process works fine, but running the obfuscated files directly causes an error that would imply a circular dependency of imports (ImportError: cannot import name 'x' from 'y'). Running the unobfuscated files directly works fine, it's only the obfuscated files that seem to have this circular dependency.

One more thing to note is without rft mode enabled, the obfuscated files run fine again without a circular dependency so this issue seems to only be tied to rft mode. Any ideas?

**Comment by jondy** [comment]
RFT mode really changes the `import` and `from...import` statement with new form, could you provide a few simple example scripts?

**Comment by chendu1993** [comment]
[test_pyarmor.zip](https://github.com/user-attachments/files/15917046/test_pyarmor.zip)
I have attached my sample scripts. What I'm running is the runsomething/runner.py
The command I'm running to obfuscate is: "pyarmor -d gen --enable-rft -O output -r ."
Running the unobfuscated version of runner.py with the correct PYTHONPATH is fine, but then running the obfuscated version in output/runsomething/runner.py with the correct PYTHONPATH is hitting an error that would imply a circular dependency:
ImportError: cannot import name 'dir2' from 'dir1.dir2.test2'

**Comment by jondy** [comment]
Got it, thanks. I'll check it when I'm back office in next Monday.

**Comment by jondy** [comment]
It will be fixed in next release, here is pre-release
https://pyarmor.dashingsoft.com/downloads/temp/pyarmor-8.5.10.zip

**Reply by chendu1993** [reply]
Hi @jondy, I'm still hitting an error with the same sample I gave. This time the error is "ImportError: cannot import name 'abcd' from 'dir1.dir2.test2'"

**Comment by jondy** [comment]
Are you really using `pyarmor -d gen --enable-rft -O output -r .` to obfuscate the scripts?
Try to use `pyarmor cfg strip_package_name = 1`, if it doesn't work, try to `pyarmor cfg strip_package_name = 2`.

**Reply by chendu1993** [reply]
I was able to verify the fix with strip_package_name = 2, thank you, but now I'm running into another similar issue with an updated sample. Here is the sample: 
[test.zip](https://github.com/user-attachments/files/15979669/test.zip)
The error I am getting now is:
ImportError: cannot import name 'Abcd' from 'dir3.test_class1'

Again, running from unobfuscated test works fine, but running from obfuscated test results in error. Thank you

**Comment by jondy** [comment]
This is because your package structure is not plain, for example, in pyarmor view, `test/test1/test3` is package `test.test1.test3`, but by extra Python path `test/test1`, it is used as package `test3`
Pyarmor RFT mode can't handle this case, it may be handled in future
But not try to reform your package structure

**Reply by chendu1993** [reply]
Makes sense, thank you for that. Is there a possible solution for the file structure to make it play well with Pyarmor RTF mode?

**Comment by jondy** [comment, marked as answer]
For example, move `test/test1/test3` to `test/test3`
If not move the folder, `import test1.test3.xxx` instead of `import test3.xxxx`
